waspc/src/Wasp/Analyzer/Evaluator/EvaluationError.hs

```diff
@@ -41,3 +55,52 @@ data EvaluationParseError
   | -- | In case when evaluation does some general parsing and it fails.
     EvaluationParseError String
   deriving (Show, Eq)
+
+mkEvaluationError :: Ctx -> EvaluationError' -> EvaluationError
+mkEvaluationError ctx e = EvaluationError $ WithCtx ctx e
+
+getErrorMessageAndCtx :: EvaluationError -> (String, Ctx)
+getErrorMessageAndCtx (EvaluationError (WithCtx ctx evalError)) = case evalError of
+  ExpectedType expectedType actualType ->
+    ( intercalate
+        "\n"
+        [ concatPrefixAndText "Expected type: " (show expectedType),
+          concatPrefixAndText "Actual type:   " (show actualType)
+        ],
+      ctx
+    )
+  ExpectedDictType actualType ->
+    ( intercalate
+        "\n"
+        [ "Expected a dictionary.",
+          concatPrefixAndText "Actual type: " (show actualType)
+        ],
+      ctx
+    )
+  ExpectedListType actualType ->
+    ( intercalate
+        "\n"
+        [ "Expected a list.",
+          concatPrefixAndText "Actual type: " (show actualType)
+        ],
+      ctx
+    )
+  UndefinedVariable varName -> ("Undefined variable " ++ varName, ctx)
+  InvalidEnumVariant enumType validEnumVariants actualEnumVariant ->
+    ( "Expected value of enum type '" ++ enumType
```

Comment:
this is awesome

Reply:
Yes, but when I finished it I realized this is one of those errors that will probably never happen due to typechecker doing most of the job and evaluations being generated automatically by our TH logic. But ok still cool :).

Reply:
And I considered doing something similar in earlier stages of compiler, but the thing is, if you type wrong identifier, e.g. `AuthhMethod`, it will fail during typechecking so early that TypeChecker doesn't yet know which type it expects there. We could probably make it work if we really wanted to, but I don't see an easy way right now to generate such nice error message for this case so I would rather leave that for later.

Reply:
Yeah, just took a look, there is certainly a way to make this work, but it would require refactoring type inference to make it more intertwined with the type "weakening", and that would take some amount of time/focus to get right, so probably best not to do it now.

Reply:
ok, a lot of cool language words here :D But I agree that we don't have to go now into it if it is complicated. We could open a gh issue for it though, so we don't forget it?

Reply:
All right, created https://github.com/wasp-lang/wasp/issues/421!